cmd/download.go

```diff
@@ -0,0 +1,185 @@
+	"reanahub/reana-client-go/pkg/workflows"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+const downloadDesc = `
+Download workspace files.
+
+The ` + "``download``" + ` command allows to download workspace files and directories.
+By default, the files specified in the workflow specification as outputs are
+downloaded. You can also specify the individual files you would like to
+download, see examples below.
+
+Examples:
+
+  $ reana-client download # download all output files
+
+  $ reana-client download mydata.tmp outputs/myplot.png
+
+  $ reana-client download -o - data.txt # write data.txt to stdout
+`
+
+const outputPathFlagDesc = `Path to the directory where files will be downloaded.
+If "-" specified as path, the files will be written to the standard output.`
+
+type downloadOptions struct {
+	token      string
+	workflow   string
+	outputPath string
+}
+
+// newDownloadCmd creates a command to download workspace files.
+func newDownloadCmd() *cobra.Command {
+	o := &downloadOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "download",
+		Short: "Download workspace files.",
+		Long:  downloadDesc,
+		Args:  cobra.ArbitraryArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return o.run(cmd, args)
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&o.token, "access-token", "t", "", "Access token of the current user.")
+	f.StringVarP(
+		&o.workflow,
+		"workflow",
+		"w", "",
+		"Name or UUID of the workflow. Overrides value of REANA_WORKON environment variable.",
+	)
+	f.StringVarP(
+		&o.outputPath,
+		"output-directory",
+		"o",
+		"",
+		outputPathFlagDesc,
+	)
+
+	return cmd
+}
+
+func (o *downloadOptions) run(cmd *cobra.Command, args []string) error {
+	var downloadPaths []string
+
+	if len(args) > 0 {
+		// download files and directories specified in arguments.
+		downloadPaths = args
+	} else {
+		// download all output files and directories specified in the reana.yaml file.
+		spec, err := workflows.GetWorkflowSpecification(o.token, o.workflow)
+		if err != nil {
+			return err
+		}
+		downloadFiles := spec.Specification.Outputs.Files
+		downloadDirs := spec.Specification.Outputs.Directories
+		downloadPaths = append(downloadFiles, downloadDirs...)
+	}
+	log.Debugf("Download paths: %s", strings.Join(downloadPaths, ", "))
+
+	for _, file := range downloadPaths {
+		fileName, fileBuf, err := workflows.DownloadFile(o.token, o.workflow, file)
+		if err != nil {
+			return err
+		}
+		if o.outputPath == config.StdoutChar {
+			err := o.displayFileContent(cmd, fileName, fileBuf)
+			if err != nil {
+				return err
+			}
+		} else {
+			err := o.storeFileContent(cmd, fileName, fileBuf)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// displayFileContent writes file(s) content to the standard output.
+func (o *downloadOptions) displayFileContent(
+	cmd *cobra.Command,
+	fileName string,
+	fileBuf *bytes.Buffer,
+) error {
+	if strings.Contains(fileName, ".zip") {
```

Comment:
There is a subtle difference between reana-client-go and the python client. In particular, when downloading a zip file with `-o -`, the zip is decompressed with reana-client-go, but not with reana-client.

```console
$ reana-client download -w workflow out/test.zip -o -
[... binary data ...]                                                             
$ ./reana-client-go download -w workflow out/test.zip -o -
[... contents of the zip file ...]
``` 
We are using the `Content-Type` header in reana-client to decide whether we should decompress the file or not (see [here](https://github.com/reanahub/reana-client/blob/e7175559735eb226662c47f67ec72993d16f573d/reana_client/api/client.py#L426-L429)), because that header is set to `application/zip` only when downloading multiple files at the same time, and not when downloading a single zip file.

What do you think about this?

Reply:
In one sense you are right, current approach to always open and traverse the zip could lead to some troubles in case a file is big, but in the other hand if user specifies to display it, we should try to achieve that.
Therefore I have a mixed feelings about this, current logic in `reana-client` is to unzip multiple files, but to display zipped binary for a single file. But why we are not displaying a single file content? I think we should keep the current approach, since the zip could contain some readable information

Reply:
The original idea of this flag was to have the same behaviour as `cat` (see https://github.com/reanahub/reana-client/issues/509, even though it doesn't talk about zip files), which just prints out the content of the file, even if it is just binary data. In general, using `-o -` with binary files does not make too much sense, unless the user wants to pipe the output to some other program.

I am (weakly) in favor of keeping the behavior of `reana-client`, that is to just print the binary data of the zip file. One of the reasons is that if we decide to decompress zip files, then we should also consider other file formats (e.g. `.rar`, `.tar.gz`, etc.).

@tiborsimko what is your opinion on this?

Reply:
Yes, I agree keeping the behaviour of `reana-client`, i.e. simply printing content regardless of its nature, to keep the analogy with `cat`.

Reply:
thank you both for your input, I've modified the code accordingly